### PR TITLE
[SHARE-482][Feature] Allow readonly admins

### PR DESCRIPTION
**Pull request by leb2dg**
### Admin addition
*`readonly` group added to staging and prod; doesn't impact anything until users are added to the group*
* create group `readonly` with the below permissions
* add the limited user to the new `readonly` group

`readonly` users will only have access to provider registrations

```
share | provider registration | Can add provider registration
share | provider registration | Can change provider registration
```

### CODE REVIEW

File: share/readonlyadmin.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        if request.method == 'POST':
+            raise PermissionDenied
+        request.readonly = True
+        return super(ReadOnlyAdmin, self).change_view(
+            request, object_id, extra_context=extra_context)
+
+    def __user_is_readonly(self, request):
+        groups = [x.name for x in request.user.groups.all()]
+
+        return "readonly" in groups
```
Comment (chrisseto) [inline share/readonlyadmin.py]: single quotes

File: share/readonlyadmin.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    def change_view(self, request, object_id, extra_context=None):
+
+        if self.__user_is_readonly(request):
+            self.readonly_fields = self.user_readonly
+            self.inlines = self.user_readonly_inlines
+
+        try:
+            return super(ReadOnlyAdmin, self).change_view(
+                request, object_id, extra_context=extra_context)
```
Comment (chrisseto) [inline share/readonlyadmin.py]: Don't wrap lines this short.

File: share/readonlyadmin.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        except PermissionDenied:
+            pass
+        if request.method == 'POST':
+            raise PermissionDenied
+        request.readonly = True
+        return super(ReadOnlyAdmin, self).change_view(
```
Comment (chrisseto) [inline share/readonlyadmin.py]: Don't wrap lines this short.

File: share/readonlyadmin.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
+## adopted from http://mike.hostetlerhome.com/blog/2012/11/13/add-a-read-only-role-to-django-admin/
+from django.contrib import admin
+from django.core.exceptions import PermissionDenied
+
+
+class ReadOnlyAdmin(admin.ModelAdmin):
+    """
```
Comment (chrisseto) [inline share/readonlyadmin.py]: Remove empty docstrings or fill them out, please.

File: share/readonlyadmin.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        except PermissionDenied:
+            pass
+        if request.method == 'POST':
+            raise PermissionDenied
+        request.readonly = True
+        return super(ReadOnlyAdmin, self).change_view(
+            request, object_id, extra_context=extra_context)
+
+    def __user_is_readonly(self, request):
+        groups = [x.name for x in request.user.groups.all()]
```
Comment (chrisseto) [inline share/readonlyadmin.py]: Not that it matters that much. You only have to iterate over `request.user.groups.all()` once.
Better yet, if you're just pulling name you can use `values_list`. 👍

File: share/readonlyadmin.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+        if self.__user_is_readonly(request):
+            if 'delete_selected' in actions:
+                del actions['delete_selected']
+
+        return actions
+
+    def change_view(self, request, object_id, extra_context=None):
+
+        if self.__user_is_readonly(request):
```
Comment (chrisseto) [inline share/readonlyadmin.py]: If you pluck this out, you won't have to give redefine all the fields on each admin model.
I don't think there are any fields that we'll be hiding.
That only use the attribute if its defined.

File: share/readonlyadmin.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            self.readonly_fields = self.user_readonly
+            self.inlines = self.user_readonly_inlines
```
Comment (chrisseto) [inline share/readonlyadmin.py]: Same as above, this one seems to be empty the majority of the time.

**Comment by leb2dg** [comment]
Gross. Sorry! I have a bad habit of not reading code that I get from other places...

File: share/readonlyadmin.py (hunk 6 above)
Comment (leb2dg) [inline share/readonlyadmin.py]: @chrisseto, I'm not following what you mean here.  Are you saying that `self.readonly_fields = self.user_readonly` should be something like `self.readonly_fields = self.model._meta.get_all_field_names()`*

*not actually correct

Comment (chrisseto) [inline share/readonlyadmin.py]: Ah, whoops. I think misunderstood something.
For the moment let's assume that there will never be a point in time that we want to give readonly users access to some fields.
You can save a lot of typing and grief if we every update fields by just getting a list of all fields dynamically, your latter suggestion/example.
Take a look at [this](http://stackoverflow.com/questions/13817525/django-admin-make-all-fields-readonly), `self.model._meta.get_all_field_names()` may not be enough.

Alternatively you could just make `readonly_fields` a "yes-man" list. From what I saw in the django code its only used for `x in self.readonly_fields`.

```python
class SetOfEverything(list):
    def __contains__(self, other):
        return True

'foo' in SetOfEverything()  # True
{} in SetOfEverything()  # True
SetOfEverything()  in SetOfEverything()  # Guess that answers that question
```

Comment (leb2dg) [inline share/readonlyadmin.py]: I'm getting the below error when programmatically including all of the fields. It only occurs when viewing the details of a celery task.
```
Internal Server Error: /admin/share/celerytask/1723/change/
Traceback (most recent call last):
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/contrib/admin/options.py", line 639, in get_form
    return modelform_factory(self.model, **defaults)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/forms/models.py", line 545, in modelform_factory
    return type(form)(class_name, (form,), form_class_attrs)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/forms/models.py", line 257, in __new__
    raise FieldError(message)
django.core.exceptions.FieldError: Unknown field(s) (traceback, timestamp) specified for CeleryTask

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/contrib/admin/options.py", line 541, in wrapper
    return self.admin_site.admin_view(view)(*args, **kwargs)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/utils/decorators.py", line 149, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/views/decorators/cache.py", line 57, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/contrib/admin/sites.py", line 244, in inner
    return view(request, *args, **kwargs)
  File "/Users/laurenbarker/GitHub/COS/SHARE/share/readonlyadmin.py", line 32, in change_view
    return super(ReadOnlyAdmin, self).change_view(request, object_id, extra_context=extra_context)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/contrib/admin/options.py", line 1440, in change_view
    return self.changeform_view(request, object_id, form_url, extra_context)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/utils/decorators.py", line 67, in _wrapper
    return bound_func(*args, **kwargs)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/utils/decorators.py", line 149, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/utils/decorators.py", line 63, in bound_func
    return func.__get__(self, type(self))(*args2, **kwargs2)
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/contextlib.py", line 30, in inner
    return func(*args, **kwds)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/contrib/admin/options.py", line 1367, in changeform_view
    ModelForm = self.get_form(request, obj)
  File "/Users/laurenbarker/.virtualenvs/share/lib/python3.5/site-packages/django/contrib/admin/options.py", line 642, in get_form
    % (e, self.__class__.__name__))
django.core.exceptions.FieldError: Unknown field(s) (traceback, timestamp) specified for CeleryTask. Check fields/fieldsets/exclude attributes of class CeleryTaskAdmin
```

File: share/admin.py [hunk 8]
```diff
@@ -154,25 +155,25 @@ def num_contributors(self, obj):
...
+class RawDataAdmin(ReadOnlyAdmin):
     raw_id_fields = ('tasks',)
 
 
-class AccessTokenAdmin(admin.ModelAdmin):
+class AccessTokenAdmin(ReadOnlyAdmin):
     list_display = ('token', 'user', 'scope')
 
 
-class ProviderRegistrationAdmin(admin.ModelAdmin):
```
Comment (chrisseto) [inline share/admin.py]: Just to have minimal changes, could you only change the base of this class?

**Pull request merged**